pinot-core/src/main/java/org/apache/pinot/core/plan/SelectionPlanNode.java

```diff
@@ -50,67 +56,152 @@ public Operator<SelectionResultsBlock> run() {
+   * considered sorted and its ASC/DESC is ignored.
+   *
+   * @return the max number that guarantees that from the first expression to the returned number, the index is already
+   * sorted.
    */
-  private boolean isAllOrderByColumnsSorted(List<OrderByExpressionContext> orderByExpressions) {
-    for (OrderByExpressionContext orderByExpression : orderByExpressions) {
-      if (!(orderByExpression.getExpression().getType() == ExpressionContext.Type.IDENTIFIER)
-          || !orderByExpression.isAsc()) {
-        return false;
+  private int getSortedColumnsPrefix(List<OrderByExpressionContext> orderByExpressions, boolean isNullHandlingEnabled) {
+    boolean asc = orderByExpressions.get(0).isAsc();
+    for (int i = 0; i < orderByExpressions.size(); i++) {
+      if (!isSorted(orderByExpressions.get(i), asc, isNullHandlingEnabled)) {
+        return i;
       }
-      String column = orderByExpression.getExpression().getIdentifier();
-      if (!_indexSegment.getDataSource(column).getDataSourceMetadata().isSorted()) {
+    }
+    // If we reach here, all are sorted
+    return orderByExpressions.size();
+  }
+
+  private boolean isSorted(OrderByExpressionContext orderByExpression, boolean asc, boolean isNullHandlingEnabled) {
+    switch (orderByExpression.getExpression().getType()) {
+      case LITERAL: {
+        return true;
+      }
+      case IDENTIFIER: {
+        if (!orderByExpression.isAsc() == asc) {
+          return false;
+        }
+        String column = orderByExpression.getExpression().getIdentifier();
+        DataSource dataSource = _indexSegment.getDataSource(column);
+        // If there are null values, we cannot trust DataSourceMetadata.isSorted
+        if (isNullHandlingEnabled) {
+          NullValueVectorReader nullValueVector = dataSource.getNullValueVector();
+          if (nullValueVector != null && !nullValueVector.getNullBitmap().isEmpty()) {
+            return false;
+          }
+        }
+        return dataSource.getDataSourceMetadata().isSorted();
+      }
+      case FUNCTION: // we could optimize monotonically increasing functions
+      default: {
         return false;
       }
     }
-    return true;
+  }
+
+  public enum OrderByAlgorithm {
+    NAIVE("naive");
+
+    /**
+     * The public name of the algorithm. This name can be used by users to specify the algorithm to use. Therefore any
+     * change here must be backward compatible.
+     *
+     * @see #findByPublicName(String)
+     */
+    private final String _publicName;
+
+    OrderByAlgorithm(String publicName) {
+      _publicName = publicName;
+    }
+
+    @Nullable
+    public static OrderByAlgorithm findByPublicName(@Nullable String publicName) {
+      if (publicName == null || publicName.equals("null") || publicName.isEmpty()) {
+        return null;
+      }
+      for (OrderByAlgorithm value : OrderByAlgorithm.values()) {
```

Comment:
We usually do `OrderByAlgorithm.valueOf(name.toUpperCase())` to avoid the linear scan

Reply:
But that doesn't let us change the name of the literals or add synonymous in the future. I like this pattern more, as it is easier to maintain in the future.

But the idea of ignoring the case is nice. I'm going to add it.